crates/oxc_linter/src/rules/typescript/consistent_indexed_object_style.rs

```diff
@@ -0,0 +1,583 @@
+                            match &idx.type_annotation.type_annotation {
+                                TSType::TSTypeLiteral(lit) => {
+                                    if !self.is_index_signature {
+                                        for member in &lit.members {
+                                            if let TSSignature::TSIndexSignature(_) = member {
+                                                ctx.diagnostic(
+                                                    ConsistentIndexedObjectStyleDiagnostic(
+                                                        "record",
+                                                        "index signature",
+                                                        idx.span,
+                                                    ),
+                                                );
+                                            }
+                                        }
+                                    }
+                                }
+                                TSType::TSTypeReference(tref) => {
+                                    if !self.is_index_signature {
+                                        match &tref.type_name {
+                                            TSTypeName::IdentifierReference(iden) => {
+                                                if iden.name != decl.id.name {
+                                                    ctx.diagnostic(
+                                                        ConsistentIndexedObjectStyleDiagnostic(
+                                                            "record",
+                                                            "index signature",
+                                                            idx.span,
+                                                        ),
+                                                    );
+                                                }
+                                            }
+                                            TSTypeName::QualifiedName(_) => ctx.diagnostic(
+                                                ConsistentIndexedObjectStyleDiagnostic(
+                                                    "record",
+                                                    "index signature",
+                                                    idx.span,
+                                                ),
+                                            ),
+                                        }
+                                    }
+                                }
+                                TSType::TSUnknownKeyword(_) | TSType::TSAnyKeyword(_) => {
+                                    ctx.diagnostic(ConsistentIndexedObjectStyleDiagnostic(
+                                        "record",
+                                        "index signature",
+                                        idx.span,
+                                    ));
```

Comment:
Why do we need to diagnostic here? 🤔

Reply:
the `TSAnyKeyword` is for this test:
```ts
interface Foo {
  [key: string]: any;
}
```
you might ask, why not just error if there is an index signature, and the answer is that in some cases, even though the setting is `Record`, an index signature is allowed. For example this:
```ts
interface Foo<T> {
  [key: string]: Foo<T> | string;
}
```

Reply:
I played around wit this but could not find a way to remove the TSUnknownKeyword & TSAnyKeyword checks and still have the tests pass 🤔

Reply:
Maybe we need to add early bailout if we find a [circle](https://github.com/typescript-eslint/typescript-eslint/pull/4347/files)?

```ts
interface Foo {
	[key: string]: Foo | string;
}
```